rust-runtime/aws-smithy-http-server/src/request/request_id.rs

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Request IDs
+//!
+//! Request IDs are of two kinds:
+//!
+//! * Client Request ID: the request ID sent from a client in the request header
+//! * Server Request ID: the unique ID generated server-side for this request
+//!
+//! Both can be used at the same time and can be used as inputs in your operation handlers.
+//! To use [`ClientRequestId`], specify which headers
+//! may contain the client request ID.
+
+use std::{
+    fmt::Display,
+    task::{Context, Poll},
+};
+
+use http::request::Parts;
+use thiserror::Error;
+use tower::{Layer, Service};
+use uuid::Uuid;
+
+use crate::{body::BoxBody, response::IntoResponse};
+
+use super::{internal_server_error, FromParts};
+
+/// Opaque type for Server Request IDs.
```

Comment:
This needs a lot more docs - pretty much embedding a lot of the stuff that it's in the RFC as API docs.